### PR TITLE
Move RemoveUnusedMembers diagnostic analyzer to IDE codestyle analyze…

### PR DESCRIPTION
**Pull request by mavasani**
…r NuGet project

This is draft PR to aid/drive the design issues in https://github.com/dotnet/roslyn/issues/38480

### CODE REVIEW

File: src/CodeStyle/Core/Analyzers/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions
+{
+    internal static partial class ITypeSymbolExtensions
```
Comment (CyrusNajmabadi) [inline src/CodeStyle/Core/Analyzers/Shared/Extensions/ITypeSymbolExtensions.cs]: is this a copy, or an extraction?

File: src/CodeStyle/Core/CodeFixes/CustomCodeActions.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            public SimpleCodeAction(string title, string equivalenceKey)
+            {
+                Title = title;
+                EquivalenceKey = equivalenceKey;
+            }
+
+            public sealed override string Title { get; }
+            public sealed override string EquivalenceKey { get; }
+
+            protected override Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/CodeStyle/Core/CodeFixes/CustomCodeActions.cs]: why not reimplement this as abstract?

File: src/CodeStyle/Core/CodeFixes/CustomCodeActions.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.CodeActions
+{
+    internal static class CustomCodeActions
+    {
+        internal abstract class SimpleCodeAction : CodeAction
```
Comment (CyrusNajmabadi) [inline src/CodeStyle/Core/CodeFixes/CustomCodeActions.cs]: SimpleDocumentChangeCodeAction?

File: src/CodeStyle/Core/CodeFixes/CustomCodeActions.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+            public SolutionChangeAction(string title, Func<CancellationToken, Task<Solution>> createChangedSolution, string equivalenceKey = null)
+                : base(title, equivalenceKey)
+            {
+                _createChangedSolution = createChangedSolution;
+            }
+
+            protected override Task<Solution> GetChangedSolutionAsync(CancellationToken cancellationToken)
+            {
+                return _createChangedSolution(cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/CodeStyle/Core/CodeFixes/CustomCodeActions.cs]: nit `=>`

File: src/CodeStyle/Core/CodeFixes/DocumentBasedFixAllProvider.cs [hunk 5]
```diff
@@ -75,41 +68,28 @@ private async Task<Document> GetDocumentFixesAsync(FixAllContext fixAllContext)
...
+            foreach (var documentIdAndTask in newDocuments)
             {
-                var newDocumentRoot = await newDocuments[i].ConfigureAwait(false);
-                if (newDocumentRoot == null)
-                {
-                    continue;
-                }
-
-                solution = solution.WithDocumentSyntaxRoot(documents[i].Id, newDocumentRoot);
+                var newDocument = await documentIdAndTask.Value.ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/CodeStyle/Core/CodeFixes/DocumentBasedFixAllProvider.cs]: based on convos with some other devs, i think you first want to do `await Task.WhenAll(newDocuments.Values)...`.  That way if any particular task throws an exception you don't lose it.

File: src/CodeStyle/Core/CodeFixes/SyntaxEditorBasedFixAllProvider.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace Microsoft.CodeAnalysis.CodeFixes
+{
+    internal abstract partial class SyntaxEditorBasedCodeFixProvider : CodeFixProvider
```
Comment (CyrusNajmabadi) [inline src/CodeStyle/Core/CodeFixes/SyntaxEditorBasedFixAllProvider.cs]: copy, or move?

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 7]
```diff
@@ -12,7 +12,11 @@
 using Xunit;
 using static Roslyn.Test.Utilities.TestHelpers;
 
+#if CODE_STYLE
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: don't like this.

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 8]
```diff
@@ -1094,7 +1098,7 @@ public void M()
             using var workspace = CreateWorkspaceFromOptions(source, testParameters);
             var diagnostics = await GetDiagnosticsAsync(workspace, testParameters).ConfigureAwait(false);
             diagnostics.Verify(Diagnostic("IDE0052", "P").WithLocation(3, 17));
-            var expectedMessage = string.Format(FeaturesResources.Private_property_0_can_be_converted_to_a_method_as_its_get_accessor_is_never_invoked, "MyClass.P");
+            var expectedMessage = "Private property 'MyClass.P' can be converted to a method as its get accessor is never invoked.";
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: how can these tests works in other languages?

File: src/Features/CSharp/Portable/RemoveUnusedMembers/CSharpRemoveUnusedMembersCodeFixProvider.cs [hunk 9]
```diff
@@ -29,5 +30,9 @@ protected override void AdjustAndAddAppropriateDeclaratorsToRemove(HashSet<Field
                 AdjustAndAddAppropriateDeclaratorsToRemove(fieldDeclarator, fieldDeclarator.Declaration.Variables, declarators);
             }
         }
+
+#if CODE_STYLE
+        protected override ISymbolDeclarationService GetSymbolDeclarationService(Document document) => CSharpSymbolDeclarationService.Instance;
+#endif
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/RemoveUnusedMembers/CSharpRemoveUnusedMembersCodeFixProvider.cs]: really don't like this approach.  Why can't CODE_STYLE use the same discovery mechanism that roslyn uses?

File: src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersCodeFixProvider.cs [hunk 10]
```diff
@@ -121,7 +132,12 @@ protected static void AdjustAndAddAppropriateDeclaratorsToRemove(SyntaxNode pare
             }
         }
 
-        private class MyCodeAction : CodeAction.DocumentChangeAction
+        private class MyCodeAction
+#if CODE_STYLE
+            : CustomCodeActions.DocumentChangeAction
+#else
+            : CodeAction.DocumentChangeAction
+#endif
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersCodeFixProvider.cs]: this is really awful.

File: src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs [hunk 11]
```diff
@@ -60,7 +61,11 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
                         //
                         return ValueUsageInfo.Write;
 
+#if CODE_STYLE
+                    case IOperation iop when iop.GetType().GetInterfaces().Any(i => i.Name == "IRecursivePatternOperation"):
+#else
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs]: IMO we should not be porting until we're on a version that supports what the feature needs.  reflection here is just busted.

**Comment by mavasani** [comment]
Closing out this stale PR.

**Pull request closed without merging**